modules/core/src/main/java/org/apache/ignite/spi/ExponentialBackoffSpiTimeout.java

```diff
@@ -0,0 +1,231 @@
+/*
```

Comment:
I do not like new class and deprecation of IgniteSpiOperationTimeoutHelper. It's still used in discovery SPI. Let's postpone the refactoring and use existing class. You should implement method __public long nextTimeoutChunk(long dfltTimeout, long maxTimeout)__ where second argument will reserve not more than maxTimeout.  Call it with DFLT_START_WAIT_TIME * 2^recon, where DFLT_START_WAIT_TIME is constant defining start time for exponential backoff.

Reply:
I think IgniteSpiOperationTimeoutHelper is poorly designed, cause it mixing exception handling and timeout calculation logic. So it's doesn't conform to Solid principle. But i don't want to change this, cause it might affect discovery, we can use new class in TcpDiscovery later on. Also new implementation encapsulates backoff calculation of 4 different timeouts separately which will allow to detect node failed on discovery faster.

Reply:
I agree that IgniteSpiOperationTimeoutHelper is poorly designed, so it is worth adding a new properly working class and creating an additional ticket for discovery SPI refactoring to use the new class.